File: creational/builder/builder.cs
using System;

class BuilderPattern
{
    class PetShopPurchase
    {
        private string description { get; set; } = "Purchase contains:\n";

        public void ExpandPurchase(string description)
        {
            description = String.Format("{0}\n", description);
            this.description += description;
        }

        public string GetDescription()
        {
            return description;
        }
    }

    interface PetShopPurchaseBuilder
    {
        void AddPet(string breed);
        void AddCollar();
        void AddBowl();
        PetShopPurchase GetPurchase();
    }

    class CatShopPurchaseBuilder : PetShopPurchaseBuilder
    {
        private PetShopPurchase purchase { get; }

        public CatShopPurchaseBuilder()
        {
            purchase = new PetShopPurchase();
        }

        public void AddPet(string breed)
        {
            string description = String.Format("{0} cat", breed);
            purchase.ExpandPurchase(description);
        }

        public void AddCollar()
        {
            purchase.ExpandPurchase("cat collar");
        }

        public void AddBowl()
        {
            purchase.ExpandPurchase("little bowl");
        }

        public PetShopPurchase GetPurchase()
        {
            return purchase;
        }
    }

    static void Main(string[] args)
    {
        PetShopPurchaseBuilder builder = new CatShopPurchaseBuilder();
        builder.AddPet("Siamese");
        builder.AddCollar();
        builder.AddBowl();

        PetShopPurchase purchase = builder.GetPurchase();
        Console.WriteLine(purchase.GetDescription());
    }
}

File: creational/builder/builder.java
public class BuilderPattern {
    public static class PetShopPurchase {
        private String description = "Purchase contains:\n";

        public void expandPurchase(String description) {
            description = String.format("%s\n", description);
            this.description += description;
        }

        public String getDescription() {
            return description;
        }
    }

    public interface PetShopPurchaseBuilder {
        public void addPet(String breed);
        public void addCollar();
        public void addBowl();
        public PetShopPurchase getPurchase();
    }

    public static class CatShopPurchaseBuilder implements PetShopPurchaseBuilder {
        private PetShopPurchase purchase;

        public CatShopPurchaseBuilder() {
            purchase = new PetShopPurchase();
        }

        public void addPet(String breed) {
            String description = String.format("%s cat", breed);
            purchase.expandPurchase(description);
        }

        public void addCollar() {
            purchase.expandPurchase("cat collar");
        }

        public void addBowl() {
            purchase.expandPurchase("little bowl");
        }

        public PetShopPurchase getPurchase() {
            return purchase;
        }
    }

    public static void main(String[] args) {
        PetShopPurchaseBuilder builder = new CatShopPurchaseBuilder();
        builder.addPet("Siamese");
        builder.addCollar();
        builder.addBowl();

        PetShopPurchase purchase = builder.getPurchase();
        System.out.println(purchase.getDescription());
    }
}
